### ISSUE TITLE
Host->Container TCP connect fails after Windows Server update 10.0.14393.576

### ISSUE THREAD

**Issue by sandersaares** [docker/docker#29444]
**Description**

After applying Windows Updates to bring my Windows server up to version 10.0.14393.576, connections from the host to the container appear to fail.

**Steps to reproduce the issue:**

1. Install Windows Server 2016 and all updates (to 10.0.14393.576)
2. `docker pull microsoft/iis`
3. `docker run --detach --ip 172.16.2.3 microsoft/iis`
4. In PowerShell `wget http://172.16.2.3`

**Describe the results you received:**

`wget : Unable to connect to the remote server`

**Describe the results you expected:**

A valid web response containing the default IIS homepage.

**Additional information you deem important (e.g. issue happens only occasionally):**

Tested only using process isolation.

Worked fine on 10.0.14393.447.

Works fine on Windows 10 even with version .576.

**Output of `docker version`:**

```
Client:
 Version:      1.14.0-dev
 API version:  1.26
 Go version:   go1.7.4
 Git commit:   d1515a5
 Built:        Thu Dec 15 09:24:59 2016
 OS/Arch:      windows/amd64

Server:
 Version:      1.14.0-dev
 API version:  1.26 (minimum version 1.24)
 Go version:   go1.7.4
 Git commit:   d1515a5
 Built:        Thu Dec 15 09:24:59 2016
 OS/Arch:      windows/amd64
 Experimental: false
```

**Output of `docker info`:**

```
Containers: 1
 Running: 1
 Paused: 0
 Stopped: 0
Images: 6
Server Version: 1.14.0-dev
Storage Driver: windowsfilter
 Windows:
Logging Driver: json-file
Plugins:
 Volume: local
 Network: l2bridge l2tunnel nat null overlay transparent
Swarm: inactive
Default Isolation: process
Kernel Version: 10.0 14393 (14393.576.amd64fre.rs1_release_inmarket.161208-2252)
Operating System: Windows Server 2016 Datacenter
OSType: windows
Architecture: x86_64
CPUs: 2
Total Memory: 7 GiB
Name: buildserver-win
ID: DP6I:NSIG:KWDV:MCJT:LWTB:7S4O:BCGW:K6ME:K224:PCF7:EEMV:ILAL
Docker Root Dir: C:\ProgramData\docker
Debug Mode (client): false
Debug Mode (server): false
Username: saaresaxinom
Registry: https://index.docker.io/v1/
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**

Azure VM.

**Comment by sandersaares** [comment]
Updated steps to be more explicit about network configuration used.

**Comment by StefanScherer** [comment]
@sandersaares It's a firewall problem. Run it with a port binding at it will work.

```
docker run --detach -p 80:80 microsoft/iis
```

**Comment by sandersaares** [comment]
I confirm that adding a port binding makes it work. This might be a satisfactory temporary workaround - thanks!

**Comment by sandersaares** [comment]
Hmm, this also seems to affect container->container traffic - with no published port, no traffic gets into a container. This is turning out to be rather annoying in our deployment scenarios, which do not do any port publishing in most cases (containers talk to each other directly),

**Comment by StefanScherer** [comment]
Yes, I have seen this behavior at least since TP5 that you have to open firewall exceptions on the host.

**Comment by thaJeztah** [comment]
ping @msabansal PTAL

**Comment by msabansal** [comment]
Hairpining of container to container traffic is supported while hairpining of host to container traffic is not supported at this time in windows. 

For container to container communication to work without hairpinning manual firewall exceptions need to be added

**Comment by sandersaares** [comment]
Can you clarify this a bit further, @msabansal? The term hairpinning is unknown to me, so it is not clear what exactly this means.

It might also help if you explained exactly how the firewall configuration affects containers, as I have not found this aspect clearly documented. Certainly, containers themselves do not appear to have a firewall (the relevant PowerShell commands all return an "endpoint not found" error) and I have not seen any obvious configuration options in the host firewall rules that would define anything for container-container traffic.

Globally allowing port 80 and 443 for everything seems to make things work, so to speak, but I suspect it may just be a coincidence and/or be violating the principle of least privilege. I would appreciate it if you would outline exactly what sort of communication patterns are supported for host<->container and container<->container traffic and how I am expected to configure them.

**Comment by msabansal** [comment]
@sandersaares Ok I understand now. Hairpining in this context refers to the host rewriting data packets and putting them back for internal containers to use. Put simply if there are two containers communicating with each other on an exposed port on the same host the data packet will flow from container A to host to Container B. This is hairpining. Container to container communication is supported but host to container communication on the exposed port is not supported at this time.

When you expose a port in a container, we configure windows firewall exception for that port. The rule is applied on the container network adapter so it does follow the principle of least privilege.

We could have configured these adapters to allow traffic within the nat subnet to flow but we didn't. For your scenario I would recommend creating a custom rule with the NAT subnet and on the ports you wish to expose. This is the best workaround I can think of. 

Get-NetFirewallRule | ? {$_.Name.StartsWith("Container")} 

Will show you the firewall rules we create. This is just as an FYI. Please refrain from modifying the rules manually.

**Comment by sandersaares** [comment]
Thanks for the many details @msabansal! I came back to this topic to dig into it deeper. Creating custom rules for the NAT network seems like a sensible approach but I got hung up on what seems like a very basic subject: how do I identify the network used by containers?

For example, I create a container network with `New-ContainerNetwork nat -SubnetPrefix 172.18.0.0/24 -GatewayAddress 172.18.0.1 -Mode NAT | Out-Null`. Now, what do I tell `New-NetFirewallRule` to make a rule that applies to this network? I see an adapter listed as "vEthernet (HNS Internal NIC)" which sounds like the right one but nothing obvious tells me that it somehow is related to the "nat" container network. I fear that some workaround I create might be broken by a future patch that changes naming that I incorrectly assume is static.

In other words, what is the "right" way to create a firewall rule that applies to the container network? An example command would be most helpful!

**Comment by msabansal** [comment]
You can create a firewall rule to allow traffic in the 172.18.0.0/24 subnet for the given port. This subnet is not routable from outside.

**Comment by msabansal** [comment]
In docker's terminology this is called as ICC (inter container communication). In linux this is allowed by default but in windows we opted to not allow it.
